Update constants/magic numbers

jonbarrow opened the pull request:
Resolves #XXX

### Changes:

Draft for now as I want to go through and add all the missing values, as well as update any existing ones that need it.

There's a lot of work going on in the common protocols lib that is using magic numbers, such as:

- https://github.com/PretendoNetwork/nex-protocols-common-go/blob/5a86458671ae9d6ed8723bb996aef550304b967e/globals/utils.go#L131
- https://github.com/PretendoNetwork/nex-protocols-common-go/blob/5a86458671ae9d6ed8723bb996aef550304b967e/matchmake-extension/get_playing_session.go#L18
- https://github.com/PretendoNetwork/nex-protocols-common-go/blob/5a86458671ae9d6ed8723bb996aef550304b967e/matchmake-extension/database/get_playing_session.go#L14
- https://github.com/PretendoNetwork/nex-protocols-common-go/blob/5a86458671ae9d6ed8723bb996aef550304b967e/matchmake-extension/database/get_playing_session.go#L56

This PR aims to add all the missing magic numbers/constants/enums so that this doesn't keep happening. Nearly all the data comes from https://github.com/tech-ticks/RTDXTools/, with some RE/guess work mixed in.

I would also like to discuss the possibility of moving towards maps for these enums/flags. Go doesn't have real enums, and that does make it a little awkward to use/reference them since it means we either risk name collisions for simple names (such as `Nothing`) or we end up with very long-winded names (such as `AutoMatchmakeOptionRecordLastGIDForParticipationCheck`). So far I have opted to be inaccurate in places and ALWAYS prefix a value with its type name, even if that wasn't the case officially, so I'm not concerned about accuracy here.

By moving to maps we can potentially use code like this:

```go
AutoMatchmakeOption.None
```

Which is a bit cleaner imo.

We already treat some types this way, such as:

- https://github.com/PretendoNetwork/nex-protocols-go/blob/28beb74753b4201e9198de68331497ee6572ab80/notifications/notification_categories.go
- https://github.com/PretendoNetwork/nex-go/blob/b5858fa5b2b6513c7470118e024a3f60b8595b2c/result_codes.go

The only question then, if we do decide to do this, is what to do about the map/struct/type names. For example, do we forego the "real" names in places for code like this:

```go
type MatchmakeOption uint32

type matchmakeOptions struct {
  None                               MatchmakeOption
  RecordLastGIDForParticipationCheck MatchmakeOption
  Reserved1                          MatchmakeOption
}

var MatchmakeOptions = matchmakeOptions{
	None:                               0,
	RecordLastGIDForParticipationCheck: 1,
	Reserved1:                          2,
}
```

This works, but we'd be accessing the data as `MatchmakeOptions.None`, where `MatchmakeOptions` is not technically the "real name" since that is being used up by the type `MatchmakeOption`

Also, we should discuss moving some of the struct field types to these known types now that we have them since it would make certain code a bit cleaner due to not needing to cast when comparing to constants

- [x] I have read and agreed to the [Code of Conduct](https://github.com/PretendoNetwork/Pretendo/blob/master/.github/CODE_OF_CONDUCT.md).
- [x] I have read and complied with the [contributing guidelines](https://github.com/PretendoNetwork/Pretendo/blob/master/.github/CONTRIBUTING.md).
- [ ] What I'm implementing was an [approved issue](../issues?q=is%3Aopen+is%3Aissue+label%3Aapproved).
- [ ] I have tested all of my changes.

jonbarrow committed:
feat(messaging): add MaxStringLength and MaxBinarySize constants

jonbarrow committed:
feat(match-making): add missing constant values

jonbarrow committed:
feat(match-making): add MatchmakeGeoIpResult enum

jonbarrow committed:
chore(match-making): make MatchmakeSelectionMethod enum consistent with the official naming

jonbarrow committed:
feat(match-making): add GatheringFlags enum

jonbarrow committed:
feat(match-making): add MatchmakeSessionOption0 enum

jonbarrow committed:
feat(match-making): add MatchmakeOption enum

jonbarrow committed:
feat(match-making): add AutoMatchmakeOption enum

jonbarrow committed:
chore(match-making): add usage details to JoinMatchmakeSessionBehaviorImAlreadyJoined

jonbarrow committed:
fix(match-making): MatchmakeSessionOption0 are flags not enum

jonbarrow committed:
feat(match-making): add FindMatchmakeSessionResultOption flags

jonbarrow committed:
feat(match-making): add MatchmakeSessionModificationFlag flags

jonbarrow committed:
chore(match-making): expand MatchmakeSelectionMethod Godoc comments

jonbarrow committed:
feat(match-making): add AnybodyParticipationPolicyArgument enum

jonbarrow committed:
feat(match-making): add ParticipationPolicy types

jonbarrow committed:
fix(match-making): fix typo in MatchmakeSessionOption0None Godoc comment

jonbarrow committed:
fix(match-making): fix AnybodyParticipationPolicyArgument typos

DaniElectra commented:
All of that sounds good to me 👍

jonbarrow commented:
@DaniElectra I did some playing around with some ideas for notifications, how does this look?

```
package main

import "fmt"

type NotificationEvents uint32
type subType uint32 // Exists strictly to limit the types of values that can be passed to NotificationEvents.Build

func (ne NotificationEvents) Build(subtype ...subType) NotificationEvents {
	category := ne * 1000

	if len(subtype) == 0 {
		return category
	}

	return category + NotificationEvents(subtype[0])
}

const (
	NotificationEventsParticipationEvent NotificationEvents = 3
)

type ParticipationEvents = subType

const (
	ParticipationEventsParticipate ParticipationEvents = 1
)

func main() {
	notificationTypeA := NotificationEventsParticipationEvent
	notificationTypeB := NotificationEventsParticipationEvent.Build()
	notificationTypeC := NotificationEventsParticipationEvent.Build(ParticipationEventsParticipate)

	fmt.Println(notificationTypeA) // 3
	fmt.Println(notificationTypeB) // 3000
	fmt.Println(notificationTypeC) // 3001
}
```

I think this helps clean up some of the notification handling, though I'm a bit iffy on the name `Build`?

jonbarrow committed:
feat(notifications): add NotificationEvents types

jonbarrow committed:
feat(notifications): add NotificationEvents.Build() method

jonbarrow committed:
fix(notifications): add protocol to UpdateNotificationData mentions

jonbarrow committed:
feat(notifications): add ParticipationEvents

jonbarrow committed:
fix(notifications): add Godoc comment to NotificationEvents.Build()

jonbarrow committed:
fix(notifications): fix ParticipationEventsParticipate name

jonbarrow committed:
fix(notifications): fix ParticipationEventsParticipate comment

jonbarrow committed:
feat(notifications): add SubscriptionEvents

jonbarrow committed:
feat(notifications): remove old category/subtype handling

jonbarrow committed:
feat(match-making): add HasFlag and HasFlags methods to flag types

DaniElectra commented:
On the PR you say that:

> // NotificationEvents represents the main category of a notification type.

But then on the `Build` function the return type is also `NotificationEvents`, and the subtype gets converted to `NotificationEvents` too for this. I think the return type should just be `uint32` or something else, but the type conversions should be logical

jonbarrow committed:
fix(notifications): make NotificationEvents.Build() return types.UInt32

jonbarrow committed:
feat(match-making): change AutoMatchmakeParam to use AutoMatchmakeOption

jonbarrow commented:
@DaniElectra I've updated the return type to just be `types.UInt32` so it can be used directly on the struct it's intended for

I've also pushed the first test of setting these constants to struct fields directly. I chose `AutoMatchmakeParam` for this just because it's the simplest, but I wanted to get your feedback on it. I had to make some changes to how the `Copy` and `Equals` functions work to support this change, which breaks some consistency, but the logic I think should be safe? We aren't using pointers anymore here and our new types are real types, not structs, so we should be able to just use normal operators for simple fields? Maybe we should consider moving *all* simple fields to doing this, as it would avoid extra function hops and stuff

DaniElectra commented:
I like that yeah! We could update every enum/flags parameter like this as it would integrate better in code

> Maybe we should consider moving all simple fields to doing this, as it would avoid extra function hops and stuff

By this you mean updating *every* simple field (such as `types.UInt32` to `uint32`)?

jonbarrow commented:
> > Maybe we should consider moving all simple fields to doing this, as it would avoid extra function hops and stuff
> 
> By this you mean updating _every_ simple field (such as `types.UInt32` to `uint32`)?

No I meant like going from this:

```go
func (amp AutoMatchmakeParam) Equals(o types.RVType) bool {
	if _, ok := o.(AutoMatchmakeParam); !ok {
		return false
	}

	other := o.(AutoMatchmakeParam)

	if amp.StructureVersion != other.StructureVersion {
		return false
	}

	if !amp.SourceMatchmakeSession.Equals(other.SourceMatchmakeSession) {
		return false
	}

	if !amp.AdditionalParticipants.Equals(other.AdditionalParticipants) {
		return false
	}

	if !amp.GIDForParticipationCheck.Equals(other.GIDForParticipationCheck) {
		return false
	}

	if amp.AutoMatchmakeOption != other.AutoMatchmakeOption {
		return false
	}

	if !amp.JoinMessage.Equals(other.JoinMessage) {
		return false
	}

	if !amp.ParticipationCount.Equals(other.ParticipationCount) {
		return false
	}

	if !amp.LstSearchCriteria.Equals(other.LstSearchCriteria) {
		return false
	}

	return amp.TargetGIDs.Equals(other.TargetGIDs)
}
```

to this:

```go
func (amp AutoMatchmakeParam) Equals(o types.RVType) bool {
	if _, ok := o.(AutoMatchmakeParam); !ok {
		return false
	}

	other := o.(AutoMatchmakeParam)

	if amp.StructureVersion != other.StructureVersion {
		return false
	}

	if !amp.SourceMatchmakeSession.Equals(other.SourceMatchmakeSession) {
		return false
	}

	if !slices.Equal(amp.AdditionalParticipants, other.AdditionalParticipants) {
		return false
	}

	if amp.GIDForParticipationCheck != other.GIDForParticipationCheck {
		return false
	}

	if amp.AutoMatchmakeOption != other.AutoMatchmakeOption {
		return false
	}

	if amp.JoinMessage != other.JoinMessage {
		return false
	}

	if !amp.ParticipationCount != other.ParticipationCount {
		return false
	}

	if !slices.Equal(amp.LstSearchCriteria, other.LstSearchCriteria) {
		return false
	}

	return slices.Equal(amp.TargetGIDs, other.TargetGIDs)
}
```

The use of the `Equals` method was mostly useful before the types update, and now is only *really* useful when either dealing with a complex struct (like `SourceMatchmakeSession` here) or with an instance of a `RVType` where we don't know the underlying type at runtime

Since our simple types (UInt32, String, etc.) are all type definitions based on real, basic, types and not structs, we can make use of standard operators/methods on them in places like this

jonbarrow committed:
feat(match-making): add IsValid method to all enum types

jonbarrow commented:
@DaniElectra In `MatchmakeSessionSearchCriteria`, the `m_MatchmakeSystemType` field is handled as a string for some reason but the value is a normal integer enum. Should we change the type of `MatchmakeSessionSearchCriteria.MatchmakeSystemType` from `types.String` to `constants.MatchmakeSystemType`? This would make it *slightly* inaccurate to the NEX type, but would make for easier integrations

DaniElectra commented:
> > > Maybe we should consider moving all simple fields to doing this, as it would avoid extra function hops and stuff
> > 
> > 
> > By this you mean updating _every_ simple field (such as `types.UInt32` to `uint32`)?
> 
> No I meant like going from this:
> 
> ```go
> func (amp AutoMatchmakeParam) Equals(o types.RVType) bool {
> 	if _, ok := o.(AutoMatchmakeParam); !ok {
> 		return false
> 	}
> 
> 	other := o.(AutoMatchmakeParam)
> 
> 	if amp.StructureVersion != other.StructureVersion {
> 		return false
> 	}
> 
> 	if !amp.SourceMatchmakeSession.Equals(other.SourceMatchmakeSession) {
> 		return false
> 	}
> 
> 	if !amp.AdditionalParticipants.Equals(other.AdditionalParticipants) {
> 		return false
> 	}
> 
> 	if !amp.GIDForParticipationCheck.Equals(other.GIDForParticipationCheck) {
> 		return false
> 	}
> 
> 	if amp.AutoMatchmakeOption != other.AutoMatchmakeOption {
> 		return false
> 	}
> 
> 	if !amp.JoinMessage.Equals(other.JoinMessage) {
> 		return false
> 	}
> 
> 	if !amp.ParticipationCount.Equals(other.ParticipationCount) {
> 		return false
> 	}
> 
> 	if !amp.LstSearchCriteria.Equals(other.LstSearchCriteria) {
> 		return false
> 	}
> 
> 	return amp.TargetGIDs.Equals(other.TargetGIDs)
> }
> ```
> 
> to this:
> 
> ```go
> func (amp AutoMatchmakeParam) Equals(o types.RVType) bool {
> 	if _, ok := o.(AutoMatchmakeParam); !ok {
> 		return false
> 	}
> 
> 	other := o.(AutoMatchmakeParam)
> 
> 	if amp.StructureVersion != other.StructureVersion {
> 		return false
> 	}
> 
> 	if !amp.SourceMatchmakeSession.Equals(other.SourceMatchmakeSession) {
> 		return false
> 	}
> 
> 	if !slices.Equal(amp.AdditionalParticipants, other.AdditionalParticipants) {
> 		return false
> 	}
> 
> 	if amp.GIDForParticipationCheck != other.GIDForParticipationCheck {
> 		return false
> 	}
> 
> 	if amp.AutoMatchmakeOption != other.AutoMatchmakeOption {
> 		return false
> 	}
> 
> 	if amp.JoinMessage != other.JoinMessage {
> 		return false
> 	}
> 
> 	if !amp.ParticipationCount != other.ParticipationCount {
> 		return false
> 	}
> 
> 	if !slices.Equal(amp.LstSearchCriteria, other.LstSearchCriteria) {
> 		return false
> 	}
> 
> 	return slices.Equal(amp.TargetGIDs, other.TargetGIDs)
> }
> ```
> 
> The use of the `Equals` method was mostly useful before the types update, and now is only _really_ useful when either dealing with a complex struct (like `SourceMatchmakeSession` here) or with an instance of a `RVType` where we don't know the underlying type at runtime
> 
> Since our simple types (UInt32, String, etc.) are all type definitions based on real, basic, types and not structs, we can make use of standard operators/methods on them in places like this

Ohh okay, that sounds good to me!

> In `MatchmakeSessionSearchCriteria`, the `m_MatchmakeSystemType` field is handled as a string for some reason but the value is a normal integer enum

That's because it *is* a string on `MatchmakeSessionSearchCriteria`, so it's an optional parameter when searching

jonbarrow commented:
> > In `MatchmakeSessionSearchCriteria`, the `m_MatchmakeSystemType` field is handled as a string for some reason but the value is a normal integer enum
> 
> That's because it _is_ a string on `MatchmakeSessionSearchCriteria`, so that it's an optional parameter when searching

I see. How do you want to handle that then? That makes it rather annoying since it would be nice to be able to reuse the existing constant here, but we'd need to handle cases where it doesn't exist. `MatchmakeSystemTypeInvalid` exists but it's a "valid" type despite the name, it's not the "doesn't exist" type we'd need

Maybe we can just add a custom value to the enum that indicates if the value is missing or not?

DaniElectra commented:
I think it can remain as it is, allowing us to handle it as we consider on the common protocols. We have to parse the value somewhere, and I'd prefer to do it in the common protocols than adding a fictional enum value just to parse it here

Also important note before we forget, we don't want to add validation checks for *every* field. For example, CTGP-7 modifies the `m_MatchmakeSystemType` field (not the `m_GameMode`) on the `MatchmakeSession` to a custom "invalid" value to avoid matching with vanilla users. This worked on Nintendo servers, so I believe that behavior should remain the same on our servers too

jonbarrow committed:
chore(match-making): add PolicyArgument type

jonbarrow committed:
feat(match-making): add GatheringState type

jonbarrow committed:
chore(match-making): remove IsValid method from AnybodyParticipationPolicyArgument

since AnybodyParticipationPolicyArgument is a type ALIAS of PolicyArgument, adding a IsValid method to AnybodyParticipationPolicyArgument actually adds it to PolicyArgument, which means that any OTHER aliases of PolicyArgument cannot define their own IsValid methods

jonbarrow committed:
feat(match-making): make FindMatchmakeSessionByParticipantParam use FindMatchmakeSessionResultOption

jonbarrow committed:
feat(match-making): make Gathering uses new constants

jonbarrow committed:
feat(match-making): make JoinMatchmakeSessionParam use JoinMatchmakeSessionBehavior

jonbarrow commented:
> Also important note before we forget, we don't want to add validation checks for _every_ field. For example, CTGP-7 modifies the `m_MatchmakeSystemType` field (not the `m_GameMode`) on the `MatchmakeSession` to a custom "invalid" value to avoid matching with vanilla users. This worked on Nintendo servers, so I believe that behavior should remain the same on our servers too

I can opt to just remove the validation checks entirely if you think that would be better. We can handle all of that on the receiving end if need be?

I've already removed validation in one place in match making (not yet pushed) because I started using a real type alias which made it impossible to add multiple `IsValid` methods to subtypes, opting to just let the RMC handler do it, so maybe that should be done everywhere?

jonbarrow committed:
chore(match-making): rename GatheringStateClosed to GatheringStateLocked

jonbarrow committed:
feat(match-making): make MatchmakeSessionSearchCriteria uses new constants

jonbarrow committed:
chore(match-making): add ExtractFrom and WriteTo methods to add constants

this allows us to more easily work with these types when encoding/decoding the structs they live in

jonbarrow committed:
feat(match-making): make MatchmakeSession uses new constants

jonbarrow committed:
feat(match-making): make UpdateMatchmakeSessionParam uses new constants

DaniElectra commented:
> > Also important note before we forget, we don't want to add validation checks for _every_ field. For example, CTGP-7 modifies the `m_MatchmakeSystemType` field (not the `m_GameMode`) on the `MatchmakeSession` to a custom "invalid" value to avoid matching with vanilla users. This worked on Nintendo servers, so I believe that behavior should remain the same on our servers too
> 
> I can opt to just remove the validation checks entirely if you think that would be better. We can handle all of that on the receiving end if need be?
> 
> I've already removed validation in one place in match making (not yet pushed) because I started using a real type alias which made it impossible to add multiple `IsValid` methods to subtypes, opting to just let the RMC handler do it, so maybe that should be done everywhere?

That sounds good to me :+1:

jonbarrow committed:
chore(match-making): remove IsValid methods from constants

opt for RMC handlers to handle these checks

jonbarrow committed:
feat(match-making): make SetState use GatheringState constant

jonbarrow committed:
feat(notifications): add NotificationEvent.MapParam

jonbarrow committed:
feat(notifications): make NotificationEvent use NotificationEvents constant

jonbarrow committed:
feat(notifications): add NotificationEventsEagleAddress

jonbarrow commented:
(rebased the branch with the latest changes)

DaniElectra commented on notifications/constants/notification_events.go:
```diff
@@ -15,17 +15,33 @@ type NotificationEvents uint32
 // Takes an optional subtype. Only the first subtype defined is used.
-func (ne NotificationEvents) Build(subtype ...subType) types.UInt32 {
-	category := types.UInt32(ne * 1000)
+func (ne NotificationEvents) Build(subtype ...subType) NotificationEvents {
```
We should split the type into the categories and the full type with the subtype. Mixing these types together is counter-intuitive

jonbarrow replied:
I'm not sure what you mean? This was done because `NotificationEvent.Type` is now of type `constants.NotificationEvents`. So when we build the `NotificationEvent` instance we would do this directly:

```go
notification := notifications_types.NewNotificationEvent()
notification.Type = notifications_constants.NotificationEventsSessionLaunched.Build()
```

If it was kept as a `types.UInt32` then we would have to explicitly cast the return value every time:

```go
notification := notifications_types.NewNotificationEvent()
notification.Type = notifications_constants.NotificationEvents(notifications_constants.NotificationEventsSessionLaunched.Build())
```

I'm not sure what you mean by counter-intuitive, or where this splitting should be done? The types are already split by category and subtype, and the usage of the method seems intuitive to me?

DaniElectra replied:
My suggestion was to split the `NotificationEvents` type, `NotificationCategory` for the actual category and `NotificationEvents` for the category + subtype. The `Build` function would look like this:

```go
func (nc NotificationCategory) Build(subtype ...SubType) NotificationEvents
```

This way we wouldn't have to cast into the `notification.Type` and the separation between "category" and "category + subtype" is more intuitive

jonbarrow replied:
The only issue with that is that `NotificationEvents` is the official name of these category types https://github.com/tech-ticks/RTDXTools/blob/232a7797e01369e9c12704f58fdde65dd3ac1c32/Assets/Scripts/Stubs/Generated/Assembly-CSharp/NexPlugin/Common.cs#L35-L53, so we would have to intentionally not use the correct official name and instead repurpose the official name for our own definition

We *have* changed official names before, but that was due to technical issues (see `RatingLockPeriodDay`, which is officially named `RatingLockPeriod` but that conflicts with `RatingLockPeriod` in `RatingLockType`)

Though I do agree that that sounds more intuitive, especially since I had added these types to the `MatchmakeExtension` protocol (see https://github.com/PretendoNetwork/nex-protocols-go/pull/100/commits/2e35b3adcba3d2f5187f7210db71e8faf5401c79). I ran into a snag there because there's 3 methods that take in notification data:

- `UpdateNotificationData` - Takes in `uiType types.UInt32, uiParam1 types.UInt64, uiParam2 types.UInt64, strParam types.String`
- `GetFriendNotificationData` - Takes in `uiType types.Int32`
- `DebugNotifyEvent` - `pid types.PID, mainType types.UInt32, subType types.UInt32, param1 types.UInt64, param2 types.UInt64, stringParam types.String`

I only ended up changing `DebugNotifyEvent` to take in `mainType notifications_constants.NotificationEvents, subType notifications_constants.SubType` because the other 2 made this a bit annoying. `UpdateNotificationData` seems to expect the fully finished built type (which isn't something we have types for at all) and `GetFriendNotificationData` wants a *signed* version of this?

Renaming `NotificationEvents` to `NotificationCategory` would make handling `UpdateNotificationData` a bit better since like you said it splits the final built type from the category, but it still has the issue of "we don't have types for these built values", and it doesn't help `GetFriendNotificationData` at all

DaniElectra replied:
> `UpdateNotificationData` seems to expect the fully finished built type (which isn't something we have types for at all)

That isn't the case, `UpdateNotificationData` only uses the category as input

`GetFriendNotificationData` being signed is an issue though, huh. I guess we can just leave things as is, no big deal

jonbarrow replied:
> > `UpdateNotificationData` seems to expect the fully finished built type (which isn't something we have types for at all)
> 
> That isn't the case, `UpdateNotificationData` only uses the category as input
> 
> `GetFriendNotificationData` being signed is an issue though, huh. I guess we can just leave things as is, no big deal

That actually makes things easier then tbh. We can definitely still do this if you want, we just have to:

1. Accept that the official name for the categories type will be misused
2. Create a signed version of the categories

We *can* make the signed part a bit easier tbh by just creating a 2nd type for it. Like:

```go
package constants

type NotificationCategory uint32
type NotificationCategorySigned int32

func (nc NotificationCategory) ToSigned() NotificationCategorySigned {
    return NotificationCategorySigned(nc)
}

func (ncs NotificationCategorySigned) ToUnsigned() NotificationCategory {
    return NotificationCategory(ncs)
}

const (
	NotificationCategorySessionLaunched NotificationCategory = 2
	NotificationCategoryParticipationEvent NotificationCategory = 3
	NotificationCategoryOwnershipChangeEvent NotificationCategory = 4
	NotificationCategoryGameNotification1 NotificationCategory = 101
	NotificationCategoryGameNotification2 NotificationCategory = 102
	NotificationCategoryGameNotification3 NotificationCategory = 103
	NotificationCategoryGameNotification4 NotificationCategory = 104
	NotificationCategoryGameNotification5 NotificationCategory = 105
	NotificationCategoryGameNotification6 NotificationCategory = 106
	NotificationCategoryGameNotification7 NotificationCategory = 107
	NotificationCategoryGameNotification8 NotificationCategory = 108
	NotificationCategoryGatheringUnregistered NotificationCategory = 109
	NotificationCategoryHostChangeEvent NotificationCategory = 110
	NotificationCategoryGameNotificationLogout NotificationCategory = 111
	NotificationCategorySubscriptionEvent NotificationCategory = 112
	NotificationCategoryGameServerMaintenance NotificationCategory = 113
	NotificationCategoryMaintenanceAnnouncement NotificationCategory = 114
	NotificationCategoryServiceItemRequestCompleted NotificationCategory = 115
	NotificationCategoryRoundStarted NotificationCategory = 116
	NotificationCategoryFirstRoundReportReceived NotificationCategory = 117
	NotificationCategoryRoundSummarized NotificationCategory = 118
	NotificationCategoryMatchmakeSystemConfigurationNotification NotificationCategory = 119
	NotificationCategoryMatchmakeSessionSystemPasswordSet NotificationCategory = 120
	NotificationCategoryMatchmakeSessionSystemPasswordClear NotificationCategory = 121
	NotificationCategoryAddedToGathering NotificationCategory = 122
	NotificationCategoryUserStatusUpdatedEvent NotificationCategory = 128
	NotificationCategoryEagleAddress NotificationCategory = 200
)
```

`GetFriendNotificationData` would just take in `NotificationCategorySigned` and we could call `ToUnsigned` in the implementation handler (or just directly cast)

DaniElectra replied:
That sounds good to me 👍

jonbarrow committed:
chore(match-making): remove legacy GatheringFlags file

jonbarrow committed:
chore(subscriber): update unknown param/type names

jonbarrow committed:
feat(subscriber): add constants

jonbarrow committed:
feat(ranking): add missing ranking constants

jonbarrow committed:
feat(ranking): make ranking protocol use new constants/enums

jonbarrow committed:
feat(ranking2): add all constants

jonbarrow committed:
feat(ranking2): make ranking2 protocol use new constants/enums

jonbarrow committed:
chore(matchmake-extension): use notifications constants in DebugNotifyEvent

jonbarrow committed:
fix(datastore): GetObjectInfos takes in a types.List[types.UInt64]

jonbarrow committed:
feat(datastore): make datastore protocol use new constants/enums

jonbarrow commented:
I believe this is ready for review now. I don't THINK I missed anything? All the protocols which have known constants should have them now, I think, as well as using those constants in the protocols themselves

If I missed any constants do let me know

Also I wanted to ask: there's a couple values in `Ranking2` which don't have constants, but might benefit from having them? Such as `Ranking2CategorySetting.ResetMonth`, which seems to be a 12-bit flag set for enabling months to reset seasons on, and `Ranking2CategorySetting.ResetDay` which denotes the day of the week/month to reset seasons on

It might be useful to create bespoke types for these, despite them not having true types? Let me know what you think @DaniElectra after you go through them, I'll be happy to add any bespoke types to whatever protocol if you think it's helpful

jonbarrow commented:
Actually I lied, I missed the Nintendo variant of the notifications protocol. How do we want to handle that? We have no known names for any of those constants, just make something up I guess?

DaniElectra commented on subscriber/types/subscriber_user_status_param.go:
```diff
@@ -0,0 +1,111 @@
+// Package types implements all the types used by the Shop protocol
```
```suggestion
// Package types implements all the types used by the Subscriber protocol
```

DaniElectra commented on match-making/constants/anybody_particiation_policy_argument.go:
```diff
@@ -0,0 +1,16 @@
+package constants
+
+// AnybodyParticipationPolicyArgument seems to determine whether or not to
+// close participation when the gathering owner changes when MatchmakeSystemTypeAnybody
+// is used?
+type AnybodyParticipationPolicyArgument = PolicyArgument
```
Could this possibly be related to the `ParticipationPolicyOpenParticipation` instead of `MatchmakeSystemTypeAnybody`? Maybe this could mean that the participation policy is actually named `ParticipationPolicyAnybody`

jonbarrow replied:
In the matchmaking client there's 2 ways the game can set these policies:

- `MatchmakeSession::SetMatchmakeSystemType(MatchmakeSystemType system_type, PolicyArgument policy_arg)`
- `Gathering::SetParticipationPolicy(ParticipationPolicy policy, PolicyArgument policy_arg)`

When `MatchmakeSession::SetMatchmakeSystemType` is called and `system_type` is either `MatchmakeSystemTypeAnybody` or `MatchmakeSystemTypeFriends`, then the client internally calls `Gathering::SetParticipationPolicy(ParticipationPolicyFriendsOnly, policy_arg)`

Otherwise the client internally calls `Gathering::SetParticipationPolicy(ParticipationPolicyOpenParticipation, 0)`

At least that's how it works in Xenoblade. `ParticipationPolicyOpenParticipation` never seems to get a policy argument for some reason:

```c
void SetMatchmakeSystemType__Q3_2nn3nex16MatchmakeSessionFQ3_2nn3nex19MatchmakeSystemTypeUi(nn::nex::MatchmakeSession *this, uint system_type, uint policy_arg) {
  this->m_MatchmakeSystemType = system_type;
  if ((1 < system_type) && (system_type < 4)) {
    SetParticipationPolicy__Q3_2nn3nex9GatheringFUiT1((nn::nex::Gathering *)this, 0x62, policy_arg);
    return;
  }

  SetParticipationPolicy__Q3_2nn3nex9GatheringFUiT1((nn::nex::Gathering *)this, 8, 0);
  return;
}

void SetParticipationPolicy__Q3_2nn3nex9GatheringFUiT1(nn::nex::Gathering *this, undefined4 participation_policy, undefined4 policy_argument) {
  this->participation_policy = participation_policy;
  this->policy_argument = policy_argument;
  return;
}
```

I have no idea why it sets `ParticipationPolicyFriendsOnly` for both `MatchmakeSystemTypeAnybody` AND `MatchmakeSystemTypeFriends` though, but given the naming of `Anybody` in both `AnybodyParticipationPolicyArgument` and `MatchmakeSystemTypeAnybody`, and from what I remember in my testing, this policy argument seems directly related to `MatchmakeSystemTypeAnybody`

DaniElectra replied:
> At least that's how it works in Xenoblade. ParticipationPolicyOpenParticipation never seems to get a policy argument for some reason

hmm, I wonder if it might be a feature from the participation policy 95, which is the equivalent of policy 8 on newer versions on NEX. On Minecraft the policies are applied properly, and the friends matchmake system types use the correct policy:

```cpp
void SetMatchmakeSystemType__Q3_2nn3nex16MatchmakeSessionFQ3_2nn3nex19MatchmakeSystemTypeUi
               (int param_1,uint system_type,undefined4 policy_arg)

{
  *(uint *)(param_1 + 0x4c) = system_type;
  if (system_type != 0) {
    if (system_type == 1) {
      SetParticipationPolicy__Q3_2nn3nex9GatheringFUiT1(param_1,95,policy_arg);
      return;
    }
    if (system_type < 4) {
      SetParticipationPolicy__Q3_2nn3nex9GatheringFUiT1(param_1,98,policy_arg);
      return;
    }
  }
  SetParticipationPolicy__Q3_2nn3nex9GatheringFUiT1(param_1,96,0);
  return;
}
```

DaniElectra reviewed (commented):
I have only skimmed through the changes, but I think everything should be good

> Also I wanted to ask: there's a couple values in Ranking2 which don't have constants, but might benefit from having them? Such as Ranking2CategorySetting.ResetMonth, which seems to be a 12-bit flag set for enabling months to reset seasons on

I agree with that, this example could benefit from having a dedicated type

> and Ranking2CategorySetting.ResetDay which denotes the day of the week/month to reset seasons on

You mention day of the *week/month*, does this mean that these two cases are represented under the same field. If that is the case, then we can add a dedicated constant here too

> Actually I lied, I missed the Nintendo variant of the notifications protocol. How do we want to handle that? We have no known names for any of those constants, just make something up I guess?

I'd say to make something up yeah

jonbarrow commented:
> You mention day of the week/month, does this mean that these two cases are represented under the same field.

Correct. Depending on the reset mode used, the `resetDay` value can either correspond to a day of the week (in the weekly reset mode) or a calendar day in an enabled month (in the monthly reset mode)

There will certainly be some overlap with values (for example I believe the weekday mode uses values 0-6, whereas the calendar days use 1-28, so the "2nd of the month" constant and "Wednesday" constant will both be value 2) but I don't think this would be an issue given that you also have the context of what mode is being used

DaniElectra commented:
Also since I saw you fixed `GetObjectInfos`, reminder that `ResetRatings` also needs to be fixed

> There will certainly be some overlap with values (for example I believe the weekday mode uses values 0-6, whereas the calendar days use 1-28, so the "2nd of the month" constant and "Wednesday" constant will both be value 2) but I don't think this would be an issue given that you also have the context of what mode is being used

Yeah that seems fair enough

jonbarrow committed:
feat(notifications): split notification event and category

jonbarrow committed:
feat(notifications): create NotificationCategorySigned

jonbarrow committed:
fix(datastore): ResetRatings takes in a types.List[types.UInt64]

jonbarrow committed:
fix(subscriber): add proper Godoc comments to SubscriberUserStatusParam

jonbarrow committed:
feat(ranking2): add bespoke Ranking2ResetDay

jonbarrow committed:
feat(ranking2): add bespoke Ranking2ResetMonth

jonbarrow committed:
feat(nintendo-notifications): add NintendoNotificationEventProfile

jonbarrow committed:
feat(nintendo-notifications): add bespoke NotificationType type

jonbarrow commented:
Actually maybe this isn't ready for review yet? I just double checked and so far only the following protocols have had constants added to them:

- DataStore
- MatchMaking
- Messaging
- Notifications
- NintendoNotifications
- Ranking
- Ranking2
- Subscriber

That's only 9 of the 32 protocols we have implemented here currently. I'm sure there's others which could use these? Such as friends, stuff like the max number of friends you can have or the max length of a status message or something?

jonbarrow committed:
feat(ticket-granting): implement LoginWithContext properly

jonbarrow committed:
feat(ticket-granting): add NEX 4 login methods

jonbarrow committed:
feat(ticket-granting): add bespoke constants

jonbarrow committed:
fix(notifications): NotificationCategorySigned encoding/decoding functions need to be signed

jonbarrow committed:
feat(messaging): add MessageFlag and RecipientType constants

jonbarrow committed:
chore(ranking): add HasFlag and HasFlags methods to flag constants

jonbarrow committed:
chore(ranking2): add HasFlag and HasFlags methods to flag constants

jonbarrow committed:
chore(datastore): add HasFlag and HasFlags methods to flag constants

jonbarrow committed:
chore: move all type methods to be directly below the type

DaniElectra commented:
I have gone through every single missing protocol available here (including non-NEX) listing some constants that could be added (those with question marks mean I'm not fully certain about them since they may be impossible to know or would be unused in normal circumstances). This is a non-exhaustive list and I may have missed other constants that could be added:

- [ ] AA User
  - [ ] Some maximum for the number of entries for ApplicationInfo lists?
- [ ] Account Management
  - [ ] Some maximum for the number of entries for lists?
  - [ ] Some maximum length for strings?
  - [ ] Some maximum length for ResultRange? (On FindByNameLike)
- [ ] Debug
  - [ ] Some maximum for the number of entries for the input list of PIDs?
  - [ ] Some limit on the number of API calls returned?
- [x] Friends (3DS)
  - [x] Limit for string lengths (MAC address, serial number, comment and game mode description)
  - [x] Limit for buffer lengths (Mii, picture and application arg)
  - [x] Change bit flags
  - [x] Mii character sets
  - [x] Maximum number of friends: 100
  - [x] Limit for list sizes (probably same as max number of friends)
  - [x] Relationship types
- [x] Friends (Wii U)
  - [x] Limit for string lengths (NNID, presence message, persistent notification, friend request message, status message)
  - [x] Limit for buffer lengths (Mii, application data)
  - [x] Change bit flags
  - [x] Maximum number of friends: 100
  - [x] Limit for list sizes (probably same as max number of friends)
- [ ] Friends
  - [ ] Some maximum for the number of entries for lists?
  - [ ] Some maximum length for strings?
  - [ ] Some maximum length for ResultRange? (On GetRelationships)
- [x] Match Making
  - [x] ParticipationPolicy 4: PasswordProtected. The PolicyArgument is set to the MD5 of a key given by the user
  - [x] ParticipationPolicy 95: Probably related to AnybodyParticipationPolicyArgument
  - [x] ParticipationPolicy 96: Related to MatchmakeSystemTypeInvite
- [ ] Matchmake Referee
  - [ ] Some maximum for the number of entries for lists
  - [ ] Some maximum length for buffer on MatchmakeRefereePersonalRoundResult
  - [ ] Limit on initialRatingValue?
- [x] Messaging
  - [x] Maximum length of StrSubject: 60 (doesn't throw an error if bigger, just capped by the server)
- [ ] NAT Traversal
  - [ ] Limit for StationURL lists
- [ ] Persistent Store
  - [ ] Some maximum length for strings and buffers?
- [x] Ranking (Legacy)
  - [x] Number of scores: 2
  - [x] Constants for each known "result code". Can we categorize these?
  - [x] Limit for common data buffer. Must be exactly 0x14 on NEX 1, unknown limit on NEX 2
  - [x] Ranking order params
  - [x] Limit for uploading multiple scores
  - [x] Limit for other lists?
- [ ] Rating
  - [ ] Limit for RatingStats and values lists
  - [ ] Limit for report data, common data and token buffer
- [ ] Remote Log Device
  - [ ] Limit for log string?
- [ ] Secure Connection
  - [ ] Limit for StationURL lists: probably 2 (public and private)
  - [ ] Limit for report data and token size
- [x] Storage Manager
  - [x] Number of slots: 5 (0-indexed)
- [ ] Subscription
  - [ ] Limit for target lists and max number of subscription datas (GetActivePlayerSubscriptionData? IIRC some request fields represent number of datas to get)
  - [ ] Limit for subscription data size
  - [ ] Privacy levels?
- [ ] Utility
  - [ ] Limit for AssociateNexUniqueIdsWithMyPrincipalId

jonbarrow commented:
> I have gone through every single missing protocol available here (including non-NEX) listing some constants that could be added (those with question marks mean I'm not fully certain about them since they may be impossible to know or would be unused in normal circumstances). This is a non-exhaustive list and I may have missed other constants that could be added:
> 
> * AA User
>   
>   * Some maximum for the number of entries for ApplicationInfo lists?
> * Account Management
>   
>   * Some maximum for the number of entries for lists?
>   * Some maximum length for strings?
>   * Some maximum length for ResultRange? (On FindByNameLike)
> * Debug
>   
>   * Some maximum for the number of entries for the input list of PIDs?
>   * Some limit on the number of API calls returned?
> * Friends (3DS)
>   
>   * Limit for string lengths (MAC address, serial number, comment and game mode description)
>   * Limit for buffer lengths (Mii, picture and application arg)
>   * Change bit flags
>   * Mii character sets
>   * Maximum number of friends: 100
>   * Limit for list sizes (probably same as max number of friends)
>   * Relationship types
> * Friends (Wii U)
>   
>   * Limit for string lengths (NNID, presence message, persistent notification, friend request message, status message)
>   * Limit for buffer lengths (Mii, application data)
>   * Change bit flags
>   * Maximum number of friends: 100
>   * Limit for list sizes (probably same as max number of friends)
> * Friends
>   
>   * Some maximum for the number of entries for lists?
>   * Some maximum length for strings?
>   * Some maximum length for ResultRange? (On GetRelationships)
> * Match Making
>   
>   * ParticipationPolicy 4: PasswordProtected. The PolicyArgument is set to the MD5 of a key given by the user
>   * ParticipationPolicy 95: Probably related to AnybodyParticipationPolicyArgument
>   * ParticipationPolicy 96: Related to MatchmakeSystemTypeInvite
> * Matchmake Referee
>   
>   * Some maximum for the number of entries for lists
>   * Some maximum length for buffer on MatchmakeRefereePersonalRoundResult
>   * Limit on initialRatingValue?
> * Messaging
>   
>   * Maximum length of StrSubject: 60 (doesn't throw an error if bigger, just capped by the server)
> * NAT Traversal
>   
>   * Limit for StationURL lists
> * Persistent Store
>   
>   * Some maximum length for strings and buffers?
> * Ranking (Legacy)
>   
>   * Number of scores: 2
>   * Constants for each known "result code". Can we categorize these?
>   * Limit for common data buffer. Must be exactly 0x14 on NEX 1, unknown limit on NEX 2
>   * Ranking order params
>   * Limit for uploading multiple scores
>   * Limit for other lists?
> * Rating
>   
>   * Limit for RatingStats and values lists
>   * Limit for report data, common data and token buffer
> * Remote Log Device
>   
>   * Limit for log string?
> * Secure Connection
>   
>   * Limit for StationURL lists: probably 2 (public and private)
>   * Limit for report data and token size
> * Storage Manager
>   
>   * Number of slots: 5 (0-indexed)
> * Subscription
>   
>   * Limit for target lists and max number of subscription datas (GetActivePlayerSubscriptionData? IIRC some request fields represent number of datas to get)
>   * Limit for subscription data size
>   * Privacy levels?
> * Utility
>   
>   * Limit for AssociateNexUniqueIdsWithMyPrincipalId

Thank you for the list! I'm wondering how we should go about getting these values. Some of them like friends and such should be doable since that server is still online, but I'm unsure about the others?

Do the clients have local checks for these maybe? Or maybe we would be just need to resort to making stuff up and fixing mistakes as we go along

DaniElectra commented:
> Do the clients have local checks for these maybe? Or maybe we would be just need to resort to making stuff up and fixing mistakes as we go along

Some of them do have checks (like Messaging) but otherwise we will have to make it up

jonbarrow committed:
feat(friends-wiiu): add constants

taken mostly from wut

jonbarrow committed:
feat(friends-3ds): add constants

taken mostly from libctru

jonbarrow committed:
feat(match-making): add ParticipationPolicyPasswordProtected constant

jonbarrow committed:
feat(match-making): add ParticipationPolicyInviteOnly constant

jonbarrow committed:
feat(match-making): add FriendsOnlyParticipationPolicyArgument type

jonbarrow committed:
chore(match-making): rename ParticipationPolicyNintendoOpenParticipation to ParticipationPolicyAnybody

this brings the naming convention more in line with ParticipationPolicyFriendsOnly, making the policy name match the argument name

jonbarrow committed:
feat(messaging): add MaxStrSubjectLength

jonbarrow committed:
chore: update all "size", "length" and "index" constants to use data type int

Go uses the int data type of the return value of len() and for indexing into slices, these changes prevent the need to constantly cast the values in these contexts

DaniElectra committed:
feat(ranking/legacy): add constants

Obtained with the best of my ability based on context and guesses.

DaniElectra committed:
feat(storage-manager): add constants

DaniElectra committed:
feat(datastore): use proper name updatePassword on ResetRating

jonbarrow commented:
@DaniElectra what are your thoughts on merging this PR as-is and moving the remaining missing constants to a pinned issue? I believe we've gotten most of the low-hanging fruit, and the constants that remain are going to either be all guess work or require a large amount of research

This PR has unintentionally blocked some other developments from continuing, so it may be best to merge this as-is with a pinned issue so the missing data can be added incrementally

DaniElectra commented:
That seems good to me!

jonbarrow commented:
Awesome! Once it's been reviewed then we should be good

DaniElectra reviewed (approved):
Let's do this :D
